src/numeric-values.js

```diff
@@ -405,4 +422,405 @@ export function multiplyTypes(type1, type2) {
+  }
+}
+
+
+function last(items) {
+  return items[items.length - 1];
+}
+
+function toNAryAstNode(operatorToken, first, second) {
+  // Treat subtraction as instead being addition, with the RHS argument instead wrapped in a special "negate" node.
+  // Treat division as instead being multiplication, with the RHS argument instead wrapped in a special "invert" node.
+
+  const type = ['+','-'].includes(operatorToken.value) ? 'ADDITION' : 'MULTIPLICATION';
+  const firstValues = first.type === type ? first.values : [first];
+  const secondValues = second.type === type ? second.values : [second];
+
+  if (operatorToken.value === '-') {
+    secondValues[0] = {type: 'NEGATE', value: secondValues[0]};
+  } else if (operatorToken.value   === '/') {
+    secondValues[0] = {type: 'INVERT', value: secondValues[0]};
+  }
+  return {type, values: [...firstValues, ...secondValues]};
+}
+
+/**
+ * Convert expression to AST using the Shunting Yard Algorithm
+ * https://en.wikipedia.org/wiki/Shunting_yard_algorithm
+ * @param {(Token | CSSFunction)[]} tokens
+ * @return {null}
+ */
+function convertTokensToAST(tokens) {
+  const operatorStack = [];
+  const tree = [];
+  while (tokens.length) {
+    const token = tokens.shift();
+    if (token instanceof NumberToken || token instanceof DimensionToken || token instanceof PercentageToken ||
+      token instanceof CSSFunction || token instanceof CSSSimpleBlock || token instanceof IdentToken) {
+      tree.push(token);
+    } else if (token instanceof DelimToken && ['*', '/', '+', '-'].includes(token.value)) {
+      while (operatorStack.length &&
+      !(last(operatorStack) instanceof LeftParenthesisToken) &&
+      precedence(last(operatorStack)) > precedence(token)) {
+        const o2 = operatorStack.pop();
+        const second = tree.pop();
+        const first = tree.pop();
+        tree.push(toNAryAstNode(o2, first, second));
+      }
+      operatorStack.push(token);
+    } else if (token instanceof LeftParenthesisToken) {
+      operatorStack.push(token);
+    } else if (token instanceof RightParenthesisToken) {
+      if (!operatorStack.length) {
+        return null;
+      }
+      while (!(last(operatorStack) instanceof LeftParenthesisToken) ) {
+        const o2 = operatorStack.pop();
+        const second = tree.pop();
+        const first = tree.pop();
+        tree.push(toNAryAstNode(o2, first, second));
+      }
+      if (!(last(operatorStack) instanceof LeftParenthesisToken)) {
+        return null;
+      }
+      operatorStack.pop();
+    } else if (token instanceof WhitespaceToken) {
+      // Consume token
+    } else {
+      return null;
+    }
+  }
+  while(operatorStack.length) {
+    if (last(operatorStack) instanceof LeftParenthesisToken) {
+      return null;
+    }
+    const o2 = operatorStack.pop()
+    const second = tree.pop();
+    const first = tree.pop();
+    tree.push(toNAryAstNode(o2, first, second));
+  }
+  return tree[0];
+}
+
+function astToCSSNumericValue(node) {
```

Comment:
I would add a comment here that this is step 4 from https://drafts.css-houdini.org/css-typed-om/#reify-a-math-expression. Took me some looking up to figure this out.

Reply:
Added a comment, and realized I had to reify `e` and `pi` here: 

https://github.com/flackr/scroll-timeline/commit/ee63815768397b3bdc38235909c1505290735fe8